### ISSUE TITLE
signal.convolve produces different results when method changed

### ISSUE THREAD

**Issue by stsievert** [scipy/scipy#7217]
`convolve` can produce different results when the parameter `method` changed. In my tests, this happens when

1. one array is of an integer dtype and the other array is a floating point dtype
2. both arrays are boolean

If we run this test below on all possible numeric type pairs, we see it fails when one of the types is an integer type (one of the `int` or `uint` types).

``` python
np.random.seed(42)
n = 3000
if 'int' in type1 or 'bool' in type1:
    x1 = np.random.choice([0, 1], size=n).astype(type1)
else:
    x1 = np.random.randn(n).astype(type1)

if 'int' in type2 or 'bool' in type2:
    x2 = np.random.choice([0, 1], size=n).astype(type2)
else:
    x2 = np.random.randn(n).astype(type2)

y = {method: convolve(x1, x2, method=method) for method in ['fft', 'direct']}

try:
    assert np.allclose(y['fft'], y['direct'], atol=1e-2)
except:
    error = np.abs(y['fft'] - y['direct'])
    print(type1, type2, np.median(error), error.max())
```

Here's all the results for when `type1 in {'uint16', 'float32', 'bool'}`

``` shell
bool bool 0.0 True
float32 uint8 0.245808 0.499962
float32 int8 0.245808 0.499962
float32 uint64 0.245809319662 0.499961117748
float32 int32 0.245809319662 0.499961117748
float32 uint16 0.245808 0.499962
float32 int16 0.245808 0.499962
float32 int64 0.245809319662 0.499961117748
float32 uint32 0.245809319662 0.499961117748
int16 complex128 0.251174051449 0.499910762802
uint16 float64 0.251174051449 0.499910762802
uint16 float32 0.251171 0.499907
uint16 float16 0.250643 0.5
uint16 complex64 0.251183 0.49991
```

**Comment by larsoner** [comment]
I no longer seem to get this error on 1.2.0, closing. Feel free to reopen if I'm testing it incorrectly.

**Comment by stsievert** [comment]
I think that's fair; some of this may have been resolved in #7211 (which got merged in https://github.com/scipy/scipy/pull/7211#issuecomment-297849326).

My tests on 1.1 and 1.2.0 don't show this issue for `*int*` dtypes like shown above; when I run the (better formatted) test at https://gist.github.com/stsievert/ce8f8835040744c8d77d3a14f5d0c2b7 I only get an error for inputs of `convolve(x: float16, h: float16)` (which print `float16 float16 0.0 0.03125` in the test above).
